### DISCUSSION TITLE
waveguide_composite with other non-standard elements

### DISCUSSION THREAD

**Discussion by Qvadi** [iqm-finland/KQCircuits#67] [Q&A] [answered]
Hi,
The waveguide_composite element is a really useful building tool for complicated structures. I have a question regarding using this element:
When building the nodes, sometimes, you may want to place another element between the nodes, for eg, a two-port hanger resonator whose probe line is part of the element structure. I have ref_ports assigned to the probeline of the hanger_resonator, and when I try to insert this hanger resonator in the Node definition when building the waveguide_composite, it fails. I also tried to align it using the ports but doesnt work. 
Is there a simpler solution to achieve this? 

Eg code, I was playing with after including the relevant imports:  

nodes = [
     Node(pya.DPoint(0, 0)),
     Node(pya.DPoint(500, 0)),
     Node(pya.DPoint(500, -300)),
     Node(pya.DPoint(200, -300), face_id="2b1"),
     Node(pya.DPoint(0, -300)),
     Node(pya.DPoint(-200, -300)),
     Node(pya.DPoint(-200, -600), face_id="1t1"),
     Node(pya.DPoint(-200, -1000)),
     Node(pya.DPoint(0, -1000)),
     Node(pya.DPoint(600, -1000), HangerResonator, align=('port_pl_a', 'port_pl_b')),
]

**Comment by qpavsmi** [comment]
Hi, thanks for sticking with KQCircuits :)

To start with, could you report the stacktrace that gets printed after the failure? It should show up in KLayout either where the PCell is supposed to be or from top toolbar: `File > Log Viewer`. Example of the failure stacktrace:

![image](https://github.com/iqm-finland/KQCircuits/assets/109152229/a82a9c50-0d09-4ce0-b412-7d1a75ef8885)

It would also help us if you could share your code of the hanger_resonator, but if you don't want to share it, we will try to help as best as we can without it

**Reply by qpavsmi** [reply]
Seems that there is no issue with the construction of nodes itself. If I copy-paste your code and replace the HangerResonator with CircularCapacitor, I get no failures with below as result. So something needs to be done in HangerResonator for it to be able to be added to WaveguideComposite

![image](https://github.com/iqm-finland/KQCircuits/assets/109152229/5614c541-b42a-4c0a-9c5a-285ba849ba88)

**Reply by Qvadi** [reply]
Hi qpavsmi,
Here's the sample code I have been experimenting with to try and understand the waveguide_composite element class. Just simple tests to add it to my custom element code later. 

----
import pya

from kqcircuits.klayout_view import KLayoutView
from kqcircuits.elements.waveguide_composite import Node, WaveguideComposite
from kqcircuits.elements.hanger_resonator import HangerResonator
from kqcircuits.elements.airbridges.airbridge import Airbridge
from kqcircuits.elements.airbridge_connection import AirbridgeConnection
from kqcircuits.elements.waveguide_coplanar_taper import WaveguideCoplanarTaper
from kqcircuits.elements.finger_capacitor_square import FingerCapacitorSquare
from kqcircuits.elements.finger_capacitor_taper import FingerCapacitorTaper
from kqcircuits.elements.waveguide_coplanar_splitter import WaveguideCoplanarSplitter, t_cross_parameters
from kqcircuits.elements.flip_chip_connectors.flip_chip_connector_rf import FlipChipConnectorRf
from kqcircuits.defaults import default_faces

view = KLayoutView()

nodes = [
     Node(pya.DPoint(0, 0)),
     Node(pya.DPoint(500, 0)),
     Node(pya.DPoint(500, -300)),
     Node(pya.DPoint(200, -300), face_id="2b1"),
     Node(pya.DPoint(0, -300)),
     Node(pya.DPoint(-200, -300)),
     Node(pya.DPoint(-200, -600), face_id="1t1"),
     Node(pya.DPoint(-200, -1000)),
     Node(pya.DPoint(0, -1000)),
     Node(pya.DPoint(600, -1000))
]

----
This simple code generates this:
![image](https://github.com/iqm-finland/KQCircuits/assets/97676306/1c77ceb7-9396-46bc-b3e9-98602fd82718)

However if I add something like as a last line in the above node list: 
    Node(pya.DPoint(800, -1000), HangerResonator)

Then it fails.

I want to add the existing hanger_resonator as part of the waveguide I am constructing with waveguide_composite since the probeline of the hanger_resonator is a two-port element which should work according to the documentation. But I haven't had any luck yet

**Reply by qpavsmi** [reply]
Show me please what it shows when it fails.

And also sharing your HangerResonator code would help, unless you're using our implementation of hanger resonator?

**Reply by qpavsmi** [reply]
Ah I see that you are using our implementation of hanger resonator :D

Will look into what this needs for this to work, or if there is a bug that we need to fix

**Reply by Qvadi** [reply]
Yes, I am using the default HangerResonator code.

![image](https://github.com/iqm-finland/KQCircuits/assets/97676306/9c1289e1-1e87-4d4b-98b7-93fc5f02b409)

My node list is:

nodes = [
     Node(pya.DPoint(0, 0)),
     Node(pya.DPoint(500, 0)),
     Node(pya.DPoint(500, -300)),
     Node(pya.DPoint(200, -300), face_id="2b1"),
     Node(pya.DPoint(0, -300)),
     Node(pya.DPoint(-200, -300)),
     Node(pya.DPoint(-200, -600), face_id="1t1"),
     Node(pya.DPoint(-200, -1000)),
     Node(pya.DPoint(0, -1000)),
     Node(pya.DPoint(600, -1000)),
     Node(pya.DPoint(800,-100), HangerResonator)   
]

**Reply by Qvadi** [reply]
I think what would be really nice is an example to show how I can drop elements into the Node definition of a waveguide_composite especially using ports. I see that there is some documentation regarding this but I was not able to fix it on my end. 
Especially elements which already have a waveguide (probeline) already defined in their element class and needs to be connected to other waveguide_composite elements.

**Reply by Qvadi** [reply]
FYI, I am trying out another version of the default hanger_resonator as a custom element :). The new element works well just as is but I fail to add it as a two-port component into the waveguide_composite structure I am trying to build.
For it to work, I need to understand the general framework of adding elements into the waveguide_composite class based on ports. I see some funny things in Klayout after tinkering with the code already but still doesn't do what I want it to. Its not repeatable in some instances for me to document it here. 
I will wait for your expert comments taking the default hanger_resonator as an example.

**Comment by iqmtestd** [comment]
Hi Qvadi,

By design, WaveguideComposite Nodes can use "simple elements", i.e. those that provide `port_a` and `port_b`. `HangerResonator` qualifies in principle and your example works for me. I've slightly changed it to have a bit more space (this is why waveguide drawing usually fails) and added a  node after the resonator to show how it is embedded in the waveguide:

![image](https://github.com/iqm-finland/KQCircuits/assets/83572540/a1b00ce7-0e1d-4cd5-a8ac-288eeff3142e)

The used code:
```
nodes = [
    Node(pya.DPoint(0, 0)),
    Node(pya.DPoint(500, 0)),
    Node(pya.DPoint(500, -300)),
    Node(pya.DPoint(200, -300), face_id="2b1"),
    Node(pya.DPoint(0, -300)),
    Node(pya.DPoint(-200, -300)),
    Node(pya.DPoint(-200, -600), face_id="1t1"),
    Node(pya.DPoint(-200, -1000)),
    Node(pya.DPoint(-200,-1600), HangerResonator),
    Node(pya.DPoint(-500, -1000)),
]

view.insert_cell(WaveguideComposite, nodes=nodes)
```

That said, it is not 100% supported. See the erroneous guiding line. I have to admit that `HangerResonator` is perhaps bit more complex than originally expected. This feature was designed for "simpler" elements, like a capacitor.  This looks like a minor bug to me.

I understand that you actually want to use `port_pl_a` and `port_pl_b`. That works even better for me. Just changing the last two Nodes to:
```
    Node(pya.DPoint(-200, -1600), HangerResonator, align=('port_pl_b', 'port_pl_a')),
    Node(pya.DPoint(-500, -2000)),
```
I get an acceptable result:
![image](https://github.com/iqm-finland/KQCircuits/assets/83572540/b704f0b8-1dbc-4001-88f4-940e1a23d8c8)

**Reply by iqmtestd** [reply]
Although, while trying to double check the above code I also got an execution error and no good result.
```
  ERROR: IndexError: list index out of range in PCellDeclaration.produce
  /home/dave/.klayout_alt/KQC_master/python/kqcircuits/elements/waveguide_composite.py:633
  /home/dave/.klayout_alt/KQC_master/python/kqcircuits/elements/waveguide_composite.py:610
  /home/dave/.klayout_alt/KQC_master/python/kqcircuits/elements/waveguide_composite.py:583
  /home/dave/.klayout_alt/KQC_master/python/kqcircuits/elements/waveguide_composite.py:501
  /home/dave/.klayout_alt/KQC_master/python/kqcircuits/elements/waveguide_composite.py:465
  /home/dave/.klayout_alt/KQC_master/python/kqcircuits/elements/element.py:398
  :/built-in-pymacros/pcell_declaration_helper.lym:218
```

Apparently there is a more serious bug than I first thought. Somehow the fact that I had run `test_waveguide_composite` first affected the internal state of KLayout in such a way that it permitted a successful outcome... This should not happen.

**Reply by Qvadi** [reply]
I agree. These were the *funny things* I referred to earlier which you are also seeing. I also used the align parameter with the port names. Sometimes, I see the resonator being inserted (in wrong orientations) and if I re-run the same code, it fails.

**Reply by Qvadi** [reply]
Any suggestions to fix this?

**Reply by qpavsmi** [reply]
We are not sure ourselves what causes this, but we have identified this as a bug in our code rather than a problem with using our API. We will soon create an issue out of this and include it as part of our backlog to fix it. The issue will be listed in github in case there will be a contribution from the community that fixes this faster than we did.

**Reply by Qvadi** [reply]
Great! thank you.

**Reply by qpavsmi** [reply]
If you want to study working examples of the WaveguideComposite, you can try it out in the GUI.

1. Start KLayout GUI
2. Open the `Element library` for bottom left
3. Find `Waveguide Composite` from the list
4. Drag and drop it into the layout
5. Use the `Partial` tool above to drag the two nodes of the waveguide to stretch it, double click in the middle of the waveguide, to add more nodes, shape it any way you want.
6. Use `Edit node` tool to click one of the nodes and try out different parameters for it

Element field could be of interest to you since you can define elements to be placed at that specific node. We offer there elements that shouldn't have any problems working (i.e. HangerResonator is not there). Also try the other arguments to see what changes they cause.

After all that, if you want to see how the waveguide composition can be implemented in code, do the following:
1. Open from top toolbar `Macros > Macro development`
2. Set `Console` option below from `Ruby` to `Python`
3. In the file viewer on the left, choose `Python` panel, open `[Local - python branch] > kqcircuits_scripts > macros > export > export_cells_as_code` by double-clicking it
4. Make sure the layout you have currently open has the waveguide composition you are interested in. Then in Macro development window, press Shift+F5 (Run script from the current tab)
5. In the console you will see python code. If you copy it and paste it to a separate macro file and run it, it should generate the exact same waveguide composition.

Hope this helps!

**Reply by iqmtestd** [reply]
FYI: https://github.com/iqm-finland/KQCircuits/issues/69

**Reply by iqmtestd** [reply, marked as answer]
Hi @Qvadi,

This is fixed. Your original example (without the `align=('port_pl_a', 'port_pl_b')` part) works just fine:
![image](https://github.com/iqm-finland/KQCircuits/assets/83572540/4f3b178c-0c02-437b-8032-ab8135a7fa5f)

The `align` parameter is not needed because the default port_a and port_b is now defined for the probeline part of the element, as this is what people usually would use in a waveguide anyway.
